cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+                                       cudf::column_device_view const ring_offsets,
+                                       cudf::column_device_view const poly_points_x,
+                                       cudf::column_device_view const poly_points_y,
+                                       uint32_t *num_hits)
+{
+  __shared__ uint32_t quad_idx, poly_idx, num_point, first_pos, qpos, num_adjusted;
+
+  // assume # of points/threads no more than num_threads_per_warp * max_warps_per_block (32*32)
+  __shared__ uint16_t data[max_warps_per_block];
+
+  // assuming 1d
+  if (threadIdx.x == 0) {
+    quad_idx     = pq_quad_id[blockIdx.x];
+    poly_idx     = pq_poly_id[blockIdx.x];
+    qpos         = block_offset[blockIdx.x];
+    num_point    = block_length[blockIdx.x];
+    first_pos    = quad_offsets.element<uint32_t>(quad_idx) + qpos;
+    num_adjusted = ((num_point - 1) / num_threads_per_warp + 1) * num_threads_per_warp;
+  }
+
+  __syncthreads();
+
+  if (threadIdx.x < max_warps_per_block) { data[threadIdx.x] = 0; }
+
+  __syncthreads();
+
+  bool in_polygon   = false;
+  uint32_t tid      = first_pos + threadIdx.x;
+  uint32_t point_id = point_indices.element<uint32_t>(tid);
+
+  if (threadIdx.x < num_point) {
+    T x = point_x.element<T>(point_id);
+    T y = point_y.element<T>(point_id);
+
+    uint32_t poly_begin = poly_offsets.element<uint32_t>(poly_idx);
+    uint32_t poly_end   = poly_idx < poly_offsets.size() - 1
+                          ? poly_offsets.element<uint32_t>(poly_idx + 1)
+                          : ring_offsets.size();
+
+    for (uint32_t ring_idx = poly_begin; ring_idx < poly_end; ring_idx++)  // for each ring
+    {
+      auto ring_begin = ring_offsets.element<uint32_t>(ring_idx);
+      auto ring_end   = ring_idx < ring_offsets.size() - 1
+                        ? ring_offsets.element<uint32_t>(ring_idx + 1)
+                        : poly_points_x.size();
+      auto ring_len = ring_end - ring_begin;
+      for (auto point_idx = 0; point_idx < ring_len; ++point_idx) {
+        T x0                 = poly_points_x.element<T>(ring_begin + ((point_idx + 0) % ring_len));
+        T y0                 = poly_points_y.element<T>(ring_begin + ((point_idx + 0) % ring_len));
+        T x1                 = poly_points_x.element<T>(ring_begin + ((point_idx + 1) % ring_len));
+        T y1                 = poly_points_y.element<T>(ring_begin + ((point_idx + 1) % ring_len));
+        bool y_between_ay_by = y0 <= y && y < y1;  // is y in range [ay, by) when ay < by
+        bool y_between_by_ay = y1 <= y && y < y0;  // is y in range [by, ay) when by < ay
+        bool y_in_bounds     = y_between_ay_by || y_between_by_ay;  // is y in range [by, ay]
+        T run                = x1 - x0;
+        T rise               = y1 - y0;
+        T rise_to_point      = y - y0;
+
+        if (y_in_bounds && x < (run / rise) * rise_to_point + x0) { in_polygon = not in_polygon; }
+      }
+    }
+  }
```

Comment:
I think most of this core computation here should be in its own function for clarity / modularity.